File: files/percent.c
#include <stdio.h>
#include <string.h>

#define BETWEEN(CHR, A, B) (A <= (CHR) && (CHR) <= B)
int hex(char c) {
  if (BETWEEN(c, '0', '9'))
    return c - '0';
  else if (BETWEEN(c, 'A', 'F'))
    return c - 'A' + 10;
  else if (BETWEEN(c, 'a', 'f'))
    return c - 'a' + 10;
  else
    return -1;
}

int main(int argc, char **argv) {
  if (argc == 3) {
    if (0 == strcmp(argv[1], "encode")) {
      for (char *c = argv[2]; *c != '\0'; ++c) {
        if (BETWEEN(*c, '0', '9')
            || BETWEEN(*c, 'A', 'Z')
            || BETWEEN(*c, 'a', 'z')
            || *c == '-'
            || *c == '.'
            || *c == '_'
            || *c == '~')
          printf("%c", (int)*c);
        else
          printf("%%%.2hhX", *c);
      }
    } else if (0 == strcmp(argv[1], "decode")) {
      for (char *c = argv[2]; *c != '\0'; ++c) {
        int a, b;
        if (*c != '%')
          printf("%c", (int)*c);
        else if (c[1] == '\0' || c[2]=='\0') {
          fprintf(stderr, "Not enough characters in hex sequence %.2s.\n", c);
          return 1;
        } else if (a = hex(c[1]), b = hex(c[2]), a == -1 || b == -1) {
          fprintf(stderr, "Invalid hex sequence %.3s.\n", c);
          return 1;
        } else {
          printf("%c", a * 16 + b);
          c += 2;
        }
      }
    } else {
      fprintf(stderr, "Usage: percent [encode/decode] <string>.\n");
      return 1;
    }
  }
  printf("\n");
  return 0;
}

File: files/percent.rs
use std::io::{stdin, BufReader, Error, ErrorKind, Read, Result, Write};
fn hex(c: u8) -> Option<u8> {
	match c {
		b'0'..=b'9' => Some(c - b'0'),
		b'A'..=b'F' => Some(c - b'A' + 10),
		b'a'..=b'f' => Some(c - b'a' + 10),
		_ => None,
	}
}
enum Direction {
	Encode,
	Decode,
}
fn encode<S>(s: S) -> Result<()>
where
	S: Iterator<Item = Result<u8>>,
{
	for c in s {
		let c = c?;
		match c {
			b'0'..=b'9' | b'A'..=b'Z' | b'a'..=b'z' | b'-' | b'.' | b'_' | b'~' => {
				print!("{}", c as char)
			}
			_ => print!("%{:02X}", c),
		}
	}
	Ok(())
}
fn decode<S>(mut iter: S) -> Result<()>
where
	S: Iterator<Item = Result<u8>>,
{
	while let Some(c) = iter.next() {
		let c = c?;
		if c == b'%' {
			let a = iter.next();
			let b = iter.next();
			if let (Some(a), Some(b)) = (a, b) {
				let a = a?;
				let b = b?;
				if let (Some(a), Some(b)) = (hex(a), hex(b)) {
					std::io::stdout().write(&[a * 16 + b])?;
				} else {
					eprintln!("Invalid hex sequence {}{}.", a, b);
					return Err(Error::from(ErrorKind::InvalidData));
				}
			} else {
				eprintln!("Hex sequence interrupted by EOF.");
				return Err(Error::from(ErrorKind::UnexpectedEof));
			}
		} else {
			std::io::stdout().write(&[c])?;
		}
	}
	Ok(())
}
fn code<S>(direction: Direction, iter: S) -> Result<()>
where
	S: Iterator<Item = Result<u8>>,
{
	match direction {
		Direction::Encode => encode(iter),
		Direction::Decode => decode(iter),
	}
}
struct ByteIterator<T>(T);
impl<T> Iterator for ByteIterator<T>
where
	T: Read,
{
	type Item = Result<u8>;
	fn next(&mut self) -> Option<<Self as Iterator>::Item> {
		let mut buf = [0];
		match self.0.read(&mut buf) {
			Ok(1) => Some(Ok(buf[0])),
			Ok(_) => None,
			Err(e) => Some(Err(e)),
		}
	}
}
fn main() -> Result<()> {
	let mut args = std::env::args();
	let _ = args.next();
	let direction = match args.next().as_deref() {
		Some("encode") => Direction::Encode,
		Some("decode") => Direction::Decode,
		_ => {
			eprintln!("Usage: percent [encode/decode] <string>.");
			return Err(Error::from(ErrorKind::InvalidInput));
		}
	};
	if let Some(s) = args.next() {
		code(direction, s.bytes().map(|x| Ok(x)))?;
		print!("\n");
		Ok(())
	} else {
		code(direction, ByteIterator(BufReader::new(stdin())))
	}
}
